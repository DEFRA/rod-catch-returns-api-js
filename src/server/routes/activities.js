import {
  Activity,
  Catch,
  River,
  SmallCatch,
  SmallCatchCount,
  Submission
} from '../../entities/index.js'
import {
  createActivitySchema,
  updateActivitySchema
} from '../../schemas/activities.schema.js'
import {
  extractRiverId,
  extractSubmissionId
} from '../../utils/entity-utils.js'
import { handleNotFound, handleServerError } from '../../utils/server-utils.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'
import { mapActivityToResponse } from '../../mappers/activities.mapper.js'
import { mapCatchToResponse } from '../../mappers/catches.mapper.js'
import { mapRiverToResponse } from '../../mappers/river.mapper.js'
import { mapSmallCatchToResponse } from '../../mappers/small-catches.mapper.js'
import { mapSubmissionToResponse } from '../../mappers/submission.mapper.js'
import { sequelize } from '../../services/database.service.js'

const BASE_ACTIVITIES_URL = '/activities/{activityId}'

export default [
  {
    method: 'POST',
    path: '/activities',
    options: {
      /**
       * Create an activity for a submission in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.payload.submission - The submission id prefixed with submissions/
       *     @param {string} request.payload.daysFishedWithMandatoryRelease - The number of days fished during the mandatory release period
       *     @param {string} request.payload.daysFishedOther - The number of days fished at other times during the season
       *     @param {string} request.payload.river - The submission id prefixed with rivers/
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const {
            submission,
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            river
          } = request.payload

          const activityData = {
            submission_id: extractSubmissionId(submission),
            river_id: extractRiverId(river),
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            version: new Date()
          }

          logger.info('Creating activity with details', activityData)

          const createdActivity = await Activity.create(activityData)

          const response = mapActivityToResponse(createdActivity.toJSON())

          return h.response(response).code(StatusCodes.CREATED)
        } catch (error) {
          return handleServerError('Error creating activity', error, h)
        }
      },
      validate: {
        payload: createActivitySchema,
        options: { entity: 'Activity' }
      },
      description: 'Create an activity for a submission in the database',
      notes: 'Create an activity for a submission in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/river`,
    options: {
      /**
       * Retrieve the river associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link River}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: River,
                required: true
              }
            ]
          })

          if (!activity) {
            return handleNotFound(
              'Activity not found or has no associated river',
              h
            )
          }

          const mappedRiver = mapRiverToResponse(activity.River.toJSON())

          return h.response(mappedRiver).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching river for activity',
            error,
            h
          )
        }
      },
      description:
        'Retrieve the river associated with an activity in the database',
      notes: 'Retrieve river associated with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/smallCatches`,
    options: {
      /**
       * Retrieve the small catches associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link SmallCatch}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activityWithCatches = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: SmallCatch,
                include: [
                  {
                    association: 'counts'
                  }
                ]
              }
            ]
          })

          if (!activityWithCatches) {
            return handleNotFound(
              `Small catches not found for ${activityId}`,
              h
            )
          }

          const mappedSmallCatches = (
            activityWithCatches.SmallCatches || []
          ).map((smallCatch) => mapSmallCatchToResponse(smallCatch))

          return h
            .response({
              _embedded: {
                smallCatches: mappedSmallCatches
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching small catches', error, h)
        }
      },
      description:
        'Retrieve the small catches associated with an activity in the database',
      notes: 'Retrieve small catches with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/catches`,
    options: {
      /**
       * Retrieve the catches (salmon and large sea trout) associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Catch}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activityWithCatches = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: Catch
              }
            ]
          })

          if (!activityWithCatches) {
            return handleNotFound(`Catches not found for ${activityId}`, h)
          }

          const mappedCatches = (activityWithCatches.Catches || []).map(
            (catchEntity) => mapCatchToResponse(catchEntity)
          )

          return h
            .response({
              _embedded: {
                catches: mappedCatches
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching catches', error, h)
        }
      },
      description:
        'Retrieve the catches (salmon and large sea trout) associated with an activity in the database',
      notes:
        'Retrieve catches (salmon and large sea trout) with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Retrieve an activity by its ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId }
          })

          if (!activity) {
            return handleNotFound(`Activity not found for ID: ${activityId}`, h)
          }

          const mappedActivity = mapActivityToResponse(activity.toJSON())

          return h.response(mappedActivity).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching activity by ID', error, h)
        }
      },
      description: 'Retrieve an activity by its ID',
      notes: 'Retrieve an activity from the database by its ID',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'DELETE',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Delete an activity by ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response indicating success or failure
       */
      handler: async (request, h) => {
        const activityId = request.params.activityId

        // Begin transaction for atomic operation
        const transaction = await sequelize.transaction()

        try {
          // Find IDs of all SmallCatch records associated with the Activity
          logger.info(
            'Deleting activity with id:%s and related records',
            activityId
          )
          const smallCatchIds = (
            await SmallCatch.findAll({
              attributes: ['id'],
              where: { activity_id: activityId },
              transaction
            })
          )?.map((smallCatch) => smallCatch.id)

          // Delete associated SmallCatchCount
          await SmallCatchCount.destroy({
            where: { small_catch_id: smallCatchIds },
            transaction
          })

          // Delete associated SmallCatches
          await SmallCatch.destroy({
            where: { activity_id: activityId },
            transaction
          })

          // Delete associated Catches
          await Catch.destroy({
            where: { activity_id: activityId },
            transaction
          })

          // Delete the Activity
          const deletedCount = await Activity.destroy({
            where: { id: activityId },
            transaction
          })

          if (deletedCount === 0) {
            await transaction.rollback()
            return handleNotFound(`Activity with ID ${activityId} not found`, h)
          }

          // Commit transaction
          await transaction.commit()

          logger.info(
            'Deleted activity with id: %s and related records',
            activityId
          )
          return h.response().code(StatusCodes.NO_CONTENT)
        } catch (error) {
          await transaction.rollback()
          return handleServerError('Error deleting activity', error, h)
        }
      },
      description: 'Delete an activity by ID',
      notes: 'Deletes an activity from the database by its ID',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'PATCH',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Update a activities in the database using the activities ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The ID of the activity to update
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        const { activityId } = request.params
        const { daysFishedWithMandatoryRelease, daysFishedOther, river } =
          request.payload

        try {
          const activity = await Activity.findByPk(activityId)

          if (!activity) {
            return handleNotFound(`Activity not found for ${activityId}`, h)
          }

          const activityData = {
            river_id: river ? extractRiverId(river) : undefined,
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            version: new Date()
          }

          logger.info(
            `Updating activity ${activityId} with details`,
            activityData
          )

          // if a value is undefined, it is not updated by Sequelize
          const updatedActivity = await activity.update(activityData)

          const mappedActivity = mapActivityToResponse(updatedActivity.toJSON())

          return h.response(mappedActivity).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error updating activity', error, h)
        }
      },
      validate: {
        payload: updateActivitySchema,
        options: { entity: 'Activity' }
      },
      description: 'Update an activity',
      notes: 'Update an activity',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/submission`,
    options: {
      /**
       * Retrieve the submission associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Submission}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: Submission,
                required: true
              }
            ]
          })

          if (!activity) {
            return handleNotFound(
              'Activity not found or has no associated submission',
              h
            )
          }

          const mappedSubmission = mapSubmissionToResponse(
            activity.Submission.toJSON()
          )

          return h.response(mappedSubmission).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching submission for activity',
            error,
            h
          )
        }
      },
      description:
        'Retrieve the submission associated with an activity in the database',
      notes: 'Retrieve submission associated with an activity in the database',
      tags: ['api', 'activities']
    }
  }
]
