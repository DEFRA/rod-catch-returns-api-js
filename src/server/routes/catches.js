import {
  mapCatchToResponse,
  mapRequestToCatch
} from '../../mappers/catches.mapper.js'
import { Catch } from '../../entities/index.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'

export default [
  {
    method: 'POST',
    path: '/catches',
    options: {
      /**
       * Create a catch (salmon and large sea trout) for an activity in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.payload.activity - The activity id prefixed with activities/
       *     @param {string} request.payload.dateCaught - The ISO 8601 date the catch was recorded (e.g., "2024-06-24T00:00:00+01:00").
       *     @param {string} request.payload.species - The species ID prefixed with "species/".
       *     @param {Object} request.payload.mass - The mass of the catch.
       *     @param {number} request.payload.mass.kg - The mass in kilograms.
       *     @param {number} request.payload.mass.oz - The mass in ounces.
       *     @param {string} request.payload.mass.type - The mass type, typically "IMPERIAL" or "METRIC".
       *     @param {string} request.payload.method - The method ID prefixed with "methods/".
       *     @param {boolean} request.payload.released - Indicates if the catch was released.
       *     @param {boolean} request.payload.onlyMonthRecorded - Indicates if only the month was recorded, to allow FMT users to report on the default dates
       *     @param {boolean} request.payload.noDateRecorded - Indicates if no specific date was recorded, to allow FMT users to report on the default month
       *     @param {string} request.payload.reportingExclude - Is this entry excluded from reporting
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const catchData = mapRequestToCatch(request.payload)

          const createdCatch = await Catch.create(catchData)

          const catchResponse = mapCatchToResponse(
            request,
            createdCatch.toJSON()
          )

          return h.response(catchResponse).code(StatusCodes.CREATED)
        } catch (error) {
          logger.error('Error create catch:', error)
          return h
            .response({ error: 'Unable to create catch' })
            .code(StatusCodes.INTERNAL_SERVER_ERROR)
        }
      },
      description:
        'Create a catch (salmon and large sea trout) for an activity in the database',
      notes:
        'Create a catch (salmon and large sea trout) for an activity in the database',
      tags: ['api', 'catches']
    }
  }
]
