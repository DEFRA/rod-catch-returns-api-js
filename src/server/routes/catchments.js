import { Catchment } from '../../entities/catchment.entity.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'

export default [
  {
    method: 'GET',
    path: '/catchments',
    options: {
      /**
       * Retrieve all the catchments in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Catchment}
       */
      handler: async (_request, h) => {
        try {
          const catchments = await Catchment.findAll()
          return h
            .response({
              _embedded: {
                catchments
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          logger.error('Error fetching catchments:', error)
          return h
            .response({ error: 'Unable to fetch catchments' })
            .code(StatusCodes.INTERNAL_SERVER_ERROR)
        }
      },
      description: 'Retrieve all the catchments in the database',
      notes: 'Retrieve all the catchments in the database',
      tags: ['api', 'catchments']
    }
  }
]
