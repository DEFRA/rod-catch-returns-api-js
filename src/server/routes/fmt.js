import { authorize, signIn } from '../../services/oidc-service.js'

export default [
  {
    path: '/oidc/authorize',
    method: 'GET',
    handler: authorize,
    options: { auth: false }
  },
  {
    path: '/oidc/callback',
    method: 'POST',
    handler: signIn,
    options: { auth: false }
  }
]
