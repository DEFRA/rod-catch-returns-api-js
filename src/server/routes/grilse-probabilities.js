import {
  deleteGrilseProbabilitiesForSeasonAndGate,
  generateCsvFromGrilseProbabilities,
  getGrilseProbabilitiesBySeasonRange,
  isGrilseProbabilityExistsForSeasonAndGate,
  processGrilseProbabilities,
  validateAndParseCsvFile
} from '../../services/grilse-probabilities.service.js'
import {
  getGrilseProbabilityRequestParamSchema,
  postGrilseProbabilityRequestParamSchema,
  postGrilseProbabilityRequestQuerySchema
} from '../../schemas/grilse-probabilities.schema.js'
import { handleNotFound, handleServerError } from '../../utils/server-utils.js'
import { GrilseProbability } from '../../entities/index.js'
import { GrilseValidationError } from '../../models/grilse-validation-error.model.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'

export default [
  {
    method: 'GET',
    path: '/grilseProbabilities',
    options: {
      /**
       * Retrieve all the species in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Species}
       */
      handler: async (_request, h) => {
        try {
          const foundGrilseProbabilities = await GrilseProbability.findAll()

          const mappedGrilseProbabilities = foundGrilseProbabilities.map(
            (grilseProbability) => grilseProbability
          )

          return h
            .response({
              _embedded: {
                grilseProbabilities: mappedGrilseProbabilities
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching grilse probabilities',
            error,
            h
          )
        }
      },
      description: 'Retrieve all the grilse probabilities in the database',
      notes: 'Retrieve all the grilse probabilities in the database',
      tags: ['api', 'grilseProbabilities']
    }
  },
  {
    method: 'POST',
    path: '/reporting/reference/grilse-probabilities/{season}/{gate}',
    options: {
      /**
       * Upload a grilse probabilities csv file to the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.season - The year which the grilse probabilities file relates to
       *     @param {string} request.params.gate - The gate which the grilse probabilities file relates to
       *     @param {boolean} request.query.overwrite - A boolean to say whether the grilse probabilities for the specified season and gate should be overridden
       *     @param {string} request.payload - The csv file
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing an empty body
       */
      handler: async (request, h) => {
        try {
          const { season, gate } = request.params
          const { overwrite } = request.query

          logger.info(
            'Validating csv file with season:%s and gate:%s',
            season,
            gate
          )
          const records = await validateAndParseCsvFile(request.payload)

          const exists = await isGrilseProbabilityExistsForSeasonAndGate(
            season,
            gate
          )

          if (exists) {
            if (!overwrite) {
              logger.info(
                'Existing data found for season:%s and gate:%s',
                season,
                gate
              )
              return h
                .response({
                  message:
                    'Existing data found for the given season and gate but overwrite parameter not set'
                })
                .code(StatusCodes.CONFLICT)
            }
            logger.info(
              'Deleting existing data for season:%s and gate:%s',
              season,
              gate
            )
            await deleteGrilseProbabilitiesForSeasonAndGate(season, gate)
          }

          const grilseProbabilities = processGrilseProbabilities(
            records,
            season,
            gate
          )

          // Bulk insert records if there are valid probabilities
          if (grilseProbabilities.length > 0) {
            await GrilseProbability.bulkCreate(grilseProbabilities)
          }

          logger.info('Created data for season:%s and gate:%s', season, gate)

          return h.response().code(StatusCodes.CREATED)
        } catch (error) {
          if (error instanceof GrilseValidationError) {
            return h
              .response({
                timestamp: new Date().toISOString(),
                status: error.status,
                message: error.message,
                path: request.path,
                ...(error.error && { error: error.error }),
                ...(error.errors && { errors: error.errors })
              })
              .code(error.status)
          }
          return handleServerError(
            'Error uploading grilse probabilities file',
            error,
            h
          )
        }
      },
      validate: {
        params: postGrilseProbabilityRequestParamSchema,
        query: postGrilseProbabilityRequestQuerySchema
      },
      description: 'Upload a grilse probabilities csv file to the database',
      notes: 'Upload a grilse probabilities csv file to the database',
      tags: ['api', 'reporting']
    }
  },
  {
    method: 'GET',
    path: '/reporting/reference/grilse-probabilities/{season}',
    options: {
      /**
       * Retrieve the grilse probabilities as a csv file from the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.season - The year which the grilse probabilities relates to, either as a single year or range
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing an empty body
       */
      handler: async (request, h) => {
        try {
          const { season } = request.params

          // Split the season if it's a range, otherwise set endSeason to startSeason
          const [startSeason, endSeason] = season.includes('-')
            ? season.split('-').map(Number)
            : [Number(season), Number(season)]

          const grilseProbabilities = await getGrilseProbabilitiesBySeasonRange(
            startSeason,
            endSeason
          )

          if (grilseProbabilities.length === 0) {
            return handleNotFound(
              `Grilse probabilities not found for ${season}`,
              h
            )
          }

          const csv = generateCsvFromGrilseProbabilities(grilseProbabilities)

          return h
            .response(csv)
            .type('text/csv')
            .header(
              'Content-Disposition',
              `attachment; filename="grilse-probabilities-${season}.csv"`
            )
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error retrieving grilse probabilities file',
            error,
            h
          )
        }
      },
      validate: {
        params: getGrilseProbabilityRequestParamSchema
      },
      description:
        'Retrieve the grilse probabilities as a csv file from the database',
      notes:
        'Retrieve the grilse probabilities as a csv file from the database',
      tags: ['api', 'reporting']
    }
  }
]
