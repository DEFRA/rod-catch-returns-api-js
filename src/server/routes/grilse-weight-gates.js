import { GrilseWeightGate } from '../../entities/index.js'
import { StatusCodes } from 'http-status-codes'
import { handleServerError } from '../../utils/server-utils.js'
import { mapGrilseWeightGateToResponse } from '../../mappers/grilse-weight-gates.mapper.js'

export default [
  {
    method: 'GET',
    path: '/grilseWeightGates',
    options: {
      /**
       * Retrieve all the grilse weight gates in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link GrilseWeightGate}
       */
      handler: async (request, h) => {
        try {
          const grilseWeightGates = await GrilseWeightGate.findAll()

          const mappedGrilseWeightGates = grilseWeightGates.map(
            (grilseWeightGate) =>
              mapGrilseWeightGateToResponse(grilseWeightGate)
          )

          return h
            .response({
              _embedded: {
                grilseWeightGates: mappedGrilseWeightGates
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching grilse weight gates',
            error,
            h
          )
        }
      },
      description: 'Retrieve all the grilse weight gates in the database',
      notes: 'Retrieve all the grilse weight gates in the database',
      tags: ['api', 'grilseWeightGates']
    }
  }
]
