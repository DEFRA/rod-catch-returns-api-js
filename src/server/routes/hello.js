export default [
  {
    method: 'GET',
    path: '/hello',
    /**
     * Handler that return hello.
     *
     * @param {import('@hapi/hapi').Request request - The Hapi request object
     * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
     * @returns {import('@hapi/hapi').ResponseObject} - A response containing hello
     */
    handler: async (request, h) => {
      return h.response('Hello mate').code(200)
    }
  }
]
