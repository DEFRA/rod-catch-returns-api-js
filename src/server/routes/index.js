import Catchments from './catchments.js'
import Licences from './licences.js'
import Region from './regions.js'
import Rivers from './rivers.js'
import Species from './species.js'

import Static from './static.js'

export const apiPrefixRoutes = [].concat(Licences, Region, Rivers, Catchments, Species)

export const rootRoutes = [].concat(Static)
