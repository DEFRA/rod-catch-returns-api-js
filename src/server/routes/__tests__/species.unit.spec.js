import {
  getMockResponseToolkit,
  getServerDetails
} from '../../../test-utils/server-test-utils.js'
import { Species } from '../../../entities/index.js'
import { handleServerError } from '../../../utils/server-utils.js'
import routes from '../species.js'

jest.mock('../../../entities/index.js')
jest.mock('../../../utils/logger-utils.js')
jest.mock('../../../utils/server-utils.js')

const [
  {
    options: { handler: getAllSpeciesHandler }
  }
] = routes

const SERVER_ERROR_SYMBOL = Symbol('SERVER_ERROR')

handleServerError.mockReturnValue(SERVER_ERROR_SYMBOL)

describe('species.unit', () => {
  describe('GET /species', () => {
    const getSpeciesData = () => [
      {
        id: '1',
        name: 'Salmon',
        smallCatchMass: 0.396893,
        createdAt: '2018-11-07T10:00:00.000+0000',
        updatedAt: '2018-11-07T10:00:00.000+0000'
      },
      {
        id: '2',
        name: 'Sea Trout',
        smallCatchMass: 0.396893,
        createdAt: '2018-11-07T10:00:00.000+0000',
        updatedAt: '2018-11-07T10:00:00.000+0000'
      }
    ]

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 200 status code and the species if the call to fetch all species is successful', async () => {
      Species.findAll.mockResolvedValueOnce(getSpeciesData())

      const result = await getAllSpeciesHandler(
        getServerDetails(),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
      expect(result.statusCode).toBe(200)
    })

    it('should call handleServerError if an error occurs while fetching species', async () => {
      const error = new Error('Database error')
      Species.findAll.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await getAllSpeciesHandler(getServerDetails(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error fetching species',
        error,
        h
      )
    })

    it('should an error response if an error occurs while fetching species', async () => {
      const error = new Error('Database error')
      Species.findAll.mockRejectedValueOnce(error)

      const result = await getAllSpeciesHandler(
        getServerDetails(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })
})
