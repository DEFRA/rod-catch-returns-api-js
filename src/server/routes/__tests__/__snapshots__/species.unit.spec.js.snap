// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`species.unit GET /species should return a 200 status code and the species if the call to fetch all species is successful 1`] = `
{
  "_embedded": {
    "species": [
      {
        "_links": {
          "self": {
            "href": "http://localhost:3000/api/species/1",
          },
          "species": {
            "href": "http://localhost:3000/api/species/1",
          },
        },
        "createdAt": "2018-11-07T10:00:00.000+0000",
        "id": "1",
        "name": "Salmon",
        "smallCatchMass": 0.396893,
        "updatedAt": "2018-11-07T10:00:00.000+0000",
      },
      {
        "_links": {
          "self": {
            "href": "http://localhost:3000/api/species/2",
          },
          "species": {
            "href": "http://localhost:3000/api/species/2",
          },
        },
        "createdAt": "2018-11-07T10:00:00.000+0000",
        "id": "2",
        "name": "Sea Trout",
        "smallCatchMass": 0.396893,
        "updatedAt": "2018-11-07T10:00:00.000+0000",
      },
    ],
  },
}
`;
